### PR TITLE
[WIP] Improvements to Terminal

### PR DESCRIPTION
**Pull request by frederikvs**
Terminal emulator was one of the major reasons I have for trying out neovim. I've used conque_term before, but it was never quite what it was supposed to be.
Upon trying out neovim's :term, I encountered some minor disturbances, which I've tried to fix in this branch.

First, I've enabled folds. I suggest you give this a try with bash-preexec (https://github.com/rcaloras/bash-preexec), and in your bashrc : 

```
  [[ -f ~/.bash-preexec.sh ]] && source ~/.bash-preexec.sh
  preexec() { echo '{{{'; }
  precmd() { echo '}}}'; }
```

and foldmethod=marker set for your terminal. For me, one of the major reasons for wanting a terminal emulator in nvim are folds. This way, when a command spews a metric tonne of output, I can just fold it up. 
On insert, folds are still disabled, to avoid garbling a screen-controlling program.

Second, I noticed that when splitting and resizing, terminals could only shrink, not grow. 

Third, it annoyed me that whenever I so much as sneezed at my terminal window, the cursor would jump to the bottom and start following terminal output. When leaving Terminal-insert mode, I want my cursor to stay where it was. When I scrolled up a bunch, in one window, I want to stay there, even when giving a command in another window on the same terminal. 
When vim's idea of the cursor is at or below the terminal's idea of the cursor, the view follows terminal output. When it isn't, you've moved up, and it's assumed you'll want to stay there.
Want tho follow the output? Press `G`. Want it to stop following? a `k` should do it, or any other "up" movement :-)
This should also help for issues #2257, #2636

The code lints cleanly, and on my machine I'm getting the same test results as on the master branch. 
Test cases related to terminal behavior have been modified. 

Also, I've added a note on security & terminal to the docs.

### CODE REVIEW

**Comment by frederikvs** [comment]
I am now able to run the tests locally. It seems an "rm build ; make distclean" was required after getting the prerequisites mentioned in https://github.com/neovim/neovim/wiki/Building-Neovim

Most of the failing tests appear to be side effects of behavior I've changed, and may change again shortly, so I'll hold off on fixing the tests until I'm happy with the new behavior, and/or I've received some feedback.

**Comment by justinmk** [comment]
> First, I've enabled folds.

Interesting. There was a similar discussion about syntax highlighting. Personally I'm in favor if anything that can be enabled *without causing weird behavior in the terminal*. I do not think we should get too ambitious with the terminal unless a feature can be made totally reliable *without* complicating the core. If `:terminal` starts to have weird quirks similar to emacs `ansi-term` or `eshell` then we messed up.

> I noticed that when splitting and resizing, terminals could only shrink, not grow.

This is being discussed elsewhere and may require changes in libvterm, so make sure to look at open issues before making changes there.

> When leaving Terminal-insert mode, I want my cursor to stay where it was

In my opinion the  terminal cursor should disappear completely. The presence of two cursors is confusing.

**Comment by frederikvs** [comment]
Thanks for your feedback. 

> Personally I'm in favor if anything that can be enabled without causing weird behavior in the terminal.

I don't think the folds cause weird behavior towards the terminal. They only modify how the terminal buffer is shown, but as far as libvterm is concerned, they don't exist. Libvterm tells us to modify certain lines of the buffer, and whether or not they're in a fold doesn't matter.
Towards the user, your mileage may vary, depending on what your terminal prints (in the way of fold-markers), and how you've configured your folds (in my case : foldmethod=marker). Default behavior will remain the same : no folds, so nothing should break. But users such as myself who want folds in the terminal can now enable them.

Could you elaborate on the quirks of emacs ansi-term and eshell? I've never worked with those myself.

> This is being discussed elsewhere and may require changes in libvterm, so make sure to look at open issues before making changes there.

The issue that I noticed does not require changes in libvterm. Before my change, on a terminal resize, the new size was the minimum of all open windows on that buffer and the current size of the terminal. With my changes, the new size is simply the minimum of all open windows on that buffer.
Regarding the open issue you're talking about, would that be #2514? That is another matter entirely. That issue is about doing fancy stuff upon resize, which may indeed require libvterm-modifications. What I'm talking about is simply the current, simple resize, but to a correct new size.

To see what I'm talking about, open a terminal, split the screen a few times in various directions, close the splits so that your terminal is again as large as possible, and try entering a few commands. The terminal will only use the size that was available when it was at its smallest, even though that size is no longer relevant.


> The presence of two cursors is confusing.

Perhaps I expressed myself unclearly : I never want two cursors visible. What I meant is this : if your're in Terminal mode, and your cursor is somewhere at the top of the window (e.g. you've just cleared the terminal), and you go back to normal mode, in the master branch your Vim-cursor jumps to the bottom of the screen. I would expect Vim's cursor to continue where the terminal's cursor was.
This is actually issue #2257

**Comment by justinmk** [comment]
> But users such as myself who want folds in the terminal can now enable them.

:+1: Seems reasonable.

> Regarding the open issue you're talking about, would that be #2514

Yes.

> That is another matter entirely. 

Ok.

**Comment by frederikvs** [comment]
I've updated some test cases to the new behavior, and fixed some bugs in my commits that were revealed by other test cases (hooray for testing!)

I've added a test for the resize behavior I described. A closer inspection revealed that under certain circumstances the terminal *could* grow (and was already tested for that). It seems like a certain combination of terminal and non-terminal views is required, and the new test case tests for one situation where the old code fails and the new code succeeds.

I hope I got everything right in the tests, altough I must admit I'm not 100% sure about the caret symbol in the screen:expect. I understand it's related to the cursor, but couldn't find any documentation on the exact meaning. Mostly I just used screen:snapshot_util().


The whole thing was rebased onto master again :-)

Locally the code lints cleanly and both 'make test' and 'make unittest' give a clean bill for each commit - apart from one test that also fails on master...

File: test/functional/terminal/window_spec.lua [hunk 1]
```diff
@@ -55,19 +55,19 @@ describe('terminal window', function()
...
-    end)
+      {3:^+--  6 lines: tty ready---------------------------}|
+      ~                                                 |
+      ~                                                 |
+      ~                                                 |
+      ~                                                 |
+      ~                                                 |
+                                                        |
+      ]], {[1] = {reverse = true}, [2] = {background = 11}, [3] = {foreground = 4, background = 248}})
+  end)
```
Comment (splinterofchaos) [inline test/functional/terminal/window_spec.lua]: Should probably also have a test that it won't show folds in insert mode, or that after showing a fold then entering insert mode, the folds won't be shown. That that seems kinda weird to me...  Since I don't use folds, I'm not sure what the expected behaviour should be.

File: src/nvim/terminal.c [hunk 2]
```diff
@@ -293,11 +293,11 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
   vterm_get_size(term->vt, &curheight, &curwidth);
 
   if (!width) {
-    width = (uint16_t)curwidth;
+    width = UINT16_MAX;
   }
 
   if (!height) {
-    height = (uint16_t)curheight;
+    height = UINT16_MAX;
```
Comment (splinterofchaos) [inline src/nvim/terminal.c]: Maybe a better strategy would be to remove the `width` and `height` parameters and set them to `UINT16_MAX` unconditionally. In the four call sites of `terminal_resize()`, two pass `0` and `0`, and the other two pass the height or width of the window being resized. However, even if we know the height of *a* window being resized, we must still check every window, so it doesn't really save us anything.

File: src/nvim/terminal.c [hunk 3]
```diff
@@ -309,6 +309,14 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
     }
```
Comment (splinterofchaos) [inline src/nvim/terminal.c]: If I opened a terminal in one tab, and then opened the terminal buffer in another tab, it would seem that this will ignore the window in the other one. I think we should use `FOR_ALL_TAB_WINDOWS()` here, just ignoring the `tabpage` argument.

File: src/nvim/terminal.c [hunk 4]
```diff
@@ -907,6 +928,16 @@ static void on_refresh(Event event)
...
       continue;
     }
+    FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+      if (wp->w_buffer == term->buf) {
+        if ( wp->w_botline > wp->w_buffer->b_ml.ml_line_count ) {
+          // if last line is visible, follow the terminal
+          wp->following_terminal = true;
+        } else {
+          wp->following_terminal = false;
+        }
```
Comment (splinterofchaos) [inline src/nvim/terminal.c]: Typically, terminals only scroll when the cursor is at the last line, Also, not sure I want unfocused windows being scrolled. I might split a terminal buffer for the specific purpose of keeping the cursor at its current position as an alternative to running `$ cmd-with-log-output | less`.

Comment (frederikvs) [inline src/nvim/terminal.c]: In some situations you don't want the unfocused window to scroll, as you described. However, in some situations you do, e.g. I've started a rather long-running command (e.g. a make) and I want to make some minor changes to other files, while monitoring the output. Then I don't want to have to "jump to the terminal window, go down, jump back" every once in a while to know what's going on, I want the window to scroll.
That's why I chose this solution : if the last line of the output (probably the line where the terminal is adding its stuff) is visible in the vim window, we'll follow what's going on. If you don't want the window to scroll, ensure that the last line is not visible. 
That's behavior I've seen in a number of terminal emulators that I was very happy with :-)

Comment (splinterofchaos) [inline src/nvim/terminal.c]: I see. Well, I'd be appeased if the terminal only scrolled when the cursor sat on the last line, and preferably only if at the end, so I could split the window and just move one line up to disable scrolling. If the terminal only has one line, I could just move to the left.

Worst case, we could make it a buffer-local option whether to scroll unconditionally, but that doesn't seem like a nice workflow (`:set noscroll | sp  | *do comamnd*`).

Comment (frederikvs) [inline src/nvim/terminal.c]: I've been thinking about it some more, and I'd propose the following:
* terminal view will scroll whenever vim's cursor is at the same line or below where the terminal thinks the cursor is - i.e. usually the last non-blank line (in normal terminal/bash operation), although not always the actual last line of the buffer. Position within that line won't matter.
* scrollback should always have at least one line

This allows the following scenarios :
* you're in terminal mode, pretty much anywhere on screen (just opened a terminal, just cleared it, or way at the bottom of a command with lots of output). You go to normal mode, perhaps switch to another buffer or whatever. Terminal spews output and the screen follows.
* Same scenario, but you go to normal mode and up one line. Screen stays where it is. 
* You go to a terminal screen. You press G and you're on the bottom and following.

All three are important scenarios I think.

If the position within the line would matter, you couldn't go to follow mode as easily. If Vim's cursor would have to be at the actual bottom (blank) line to follow, and you've just cleared the terminal, entered a command and gone back to normal mode, it wouldn't follow (in the new behavior where Vim's cursor picks up where the terminal's left off).

I think/hope this would also meet your expectations, if not, please let me know.

I'll report back when I've found the time to implement this - not sure yet when :-)

Comment (splinterofchaos) [inline src/nvim/terminal.c]: If I open a new terminal buffer and run a command, the scroll back should be empty. If "scrollback should always have at least one line", doesn't that mean it won't scroll?

BTW: I realized that when you press enter, the cursor does move down one, so that means even on a new terminal window, the trick of moving up one after starting the command will still work.

But on the surface, sounds like a good plan. :+1:

**Comment by frederikvs** [comment]
After some very valuable feedback from @splinterofchaos (thanks!), I've implemented some new behavior. If you just leave terminal-insert, the window will scroll with the terminal. If you go up one line, it will stop scrolling.
All feedback remains welcome.

One weird thing popped up, and I'm not sure how to handle it.
If you do the following : 
* open a terminal
* set scrolloff=3
* get enough output to fill your screen
* for i in {1..20}; do sleep 1 ; echo "ping $i" ; done
* while that's running, go to normal mode (screen will follow terminal), and go up one line (shouldn't follow anymore).
* notice the 'weird' behavior : the screen will still scroll until you've reached your scrolloff.
This makes sense really, the terminal isn't trying to follow anymore, but each time a new line is introduced, nvim wants to get one line closer to your scrolloff.

I see 3 options, but would welcome better suggestions :
* Behavior is, strictly speaking, correct. Keep it. It will however seem a little weird/confusing to people who happen to have scrolloff set.
* Disable scrolloff in the terminal buffer. However, if somebody, for whatever reason, desires scrolloff, he'll be unhappy. 
* Make it a setting somewhere, whether scrolloff in the terminal should be disabled or not. Default should be disabled.

There's also a mixed-option : disable it for now, and when somebody comes asking for scrolloff in the terminal, make it a setting.

Stuff currently on my todo-list : 
- [ ] Remove unnecessary parameter from terminal_resize
- [ ] Add new test cases
  - [ ] term & folds
  - [ ] term & tabs
- [ ] decide on the scrolloff-behavior and implement

Anything I missed?

**Comment by frederikvs** [comment]
Of course that's simpler, where was my mind? :-)
gives some trouble with the 80-char limit though, I hope the new version is acceptable...

Fixed, rebased
One question though, regarding the space between wp and the open paren : why wasn't this caught by lint? Lint catches stuff like " '){' should be ') {' ", why not this?


I've started adding some tests, but I'm currently blocked by issue #2732. The screen states aren't what I expect them to be. That's why there's a bunch of screen:snapshot_util() calls in the tests, they're to be replaced by the proper screen:expect calls when that issue is resolved.

File: src/nvim/buffer_defs.h [hunk 5]
```diff
@@ -1158,6 +1158,12 @@ struct window_S {
    * In a non-location list window, w_llist_ref is NULL.
    */
   qf_info_T   *w_llist_ref;
+
+  /*
+   * Are we following the terminal? (Only valid if w_buffer is a terminal
+   * buffer.)
+   */
+  bool following_terminal;
```
Comment (justinmk) [inline src/nvim/buffer_defs.h]: Why not use the existing buffer-local `'autoread'` setting for this purpose?

Comment (frederikvs) [inline src/nvim/buffer_defs.h]: Indeed, we need to decide for each window whether it is following the terminal at that moment in time... A variable that's tied to the buffer is too broad.
If there are any suggestions for window-local variables that could be reused, I'll gladly switch. I had a quick look at what's already in the `struct window_S`, and could find nothing that's obviously suitable. OTOH my knowledge of those data structures is rather limited, so I could have missed something...

Comment (splinterofchaos) [inline src/nvim/buffer_defs.h]: I haven't investigated this as fully as it deserves yet, but would it make sense to write a function, `window_should_follow()` determined by its current state, or do we have to determine this at a specific point?

**Comment by frederikvs** [comment]
I've rebased everything onto the new master, added some tests, and made it work better across tabs.

I've also modified it so that even on resizes it doesn't start following. If I want it on the bottom, following, I'll press G. It's way more intuitive that way. There's no reason to jump to the bottom when resizing...
Note that this could make the "force" parameter to adjust_topline unnecessary. I haven't removed it yet because I want to hear comments first. And maybe somebody has a future use of that parameter in mind...

There does appear to be a bug somwhere, sometimes it'll stop following unexpectedly (when focus is in another window and the terminal is scrolling output). Investigating.
To reproduce : have a split screen, terminal in one. Start e.g. a `make lint`, switch to the other screen, and wait...

**Comment by frederikvs** [comment]
Hmm, I have a bug that's been annoying me for the past week, but I don't know enough about the inner workings of folds to solve it. Any help?

To reproduce :  take my `terminal` branch, compile & run it, and open a `:terminal`. Do `set fdm=marker` and run 
```
echo "{{{" ; make lint; make lint; make lint; echo "}}}" 
```
(the make lint is just to produce some output - usually it'll happen in the first make lint, but I've thrown in a few to be sure.)

As soon as you started that command, go back to normal mode, and open all folds (`zR`). At first the screen will follow the terminal output, but then it will stop. If you press `G` it'll start following again, and quite quickly stop again.

When those foldmarkers are not present, or `fdm` is set to something else, it keeps following nicely untill the end. This gives me the idea that it's probably either a bug in folding (not able to handle the lines coming in from the terminal?), or something that I'm doing wrong with folds. (Probably something I'm **not** doing...)
It's not related to the toggling of `w_p_fen` on terminal_enter - I've disabled that locally and tried again, behavior was the same.

I can reproduce it on my mac (mavericks), and on a physical ubuntu machine. Strangely enough, I can't seem to reproduce it on an ubuntu VM I have lying around. Not sure if that's a hint, or just something to confuse the hell out of me...

Any help is greatly appreciated!

**Comment by mulianov** [comment]
any updates?

File: src/nvim/terminal.c [hunk 6]
```diff
@@ -413,7 +423,16 @@ void terminal_enter(bool process_deferred)
...
   unshowmode(true);
   redraw(term->buf != curbuf);
+  if (curbuf == term->buf) {
+    curwin->w_p_fen = save_w_p_fen;
+    if (mouse_event == false) {
+      // Upon leaving Terminal mode, keep cursor where Terminal showed it
+      curwin->w_cursor.lnum = row_to_linenr(term, term->cursor.row);
+      curwin->w_cursor.col = term->cursor.col;
+    }
+  }
```
Comment (justinmk) [inline src/nvim/terminal.c]: This should be broken out into a separate PR, it's a rather high-priority improvement.

File: src/nvim/terminal.c [hunk 7]
```diff
@@ -413,7 +423,16 @@ void terminal_enter(bool process_deferred)
...
   ui_busy_stop();
+  ui_refresh();
```
Comment (justinmk) [inline src/nvim/terminal.c]: I don't see what this is for, but it might be better to do it via vimscript in a BufEnter autocmd or something like that.

**Comment by XVilka** [comment]
It should be rebased for a better testing.

**Comment by justinmk** [comment]
still interested in this but it's linked from https://github.com/neovim/neovim/issues/5431 so doesn't need to stay open
